input/kafka/inputkafka.go

```diff
@@ -0,0 +1,200 @@
+
+	saConf *sarama.Config
+}
+
+// DefaultInputConfig returns an InputConfig struct with default values
+func DefaultInputConfig() InputConfig {
+	return InputConfig{
+		InputConfig: config.InputConfig{
+			CommonConfig: config.CommonConfig{
+				Type: ModuleName,
+			},
+		},
+	}
+}
+
+// InitHandler initialize the input plugin
+func InitHandler(ctx context.Context, raw *config.ConfigRaw) (config.TypeInputConfig, error) {
+	conf := DefaultInputConfig()
+	err := config.ReflectConfig(raw, &conf)
+	if err != nil {
+		return nil, err
+	}
+
+	sarama.Logger = goglog.Logger
+
+	version, err := sarama.ParseKafkaVersion(conf.Version)
+	if err != nil {
+		goglog.Logger.Errorf("Error parsing Kafka version: %v", err)
+		return nil, err
+	}
+
+	/**
+	 * Construct a new Sarama configuration.
+	 * The Kafka cluster version has to be defined before the consumer/producer is initialized.
+	 */
+	sarConfig := sarama.NewConfig()
+	sarConfig.Version = version
+
+	switch conf.Assignor {
+	case "roundrobin":
+		sarConfig.Consumer.Group.Rebalance.Strategy = sarama.BalanceStrategyRoundRobin
+	case "range":
+		sarConfig.Consumer.Group.Rebalance.Strategy = sarama.BalanceStrategyRange
+	default:
+		goglog.Logger.Errorf("Unrecognized consumer group partition assignor: %s", conf.Assignor)
+	}
+
+	if conf.OffsetOldest {
+		sarConfig.Consumer.Offsets.Initial = sarama.OffsetOldest
+	}
+
+	if len(conf.Topics) < 0 {
+		goglog.Logger.Error("topics should not be empty")
+		return nil, err
+	}
+
+	if conf.Group == "" {
+		goglog.Logger.Error("group should not be empty")
+		return nil, err
+	}
+
+	if len(conf.Brokers) == 0 {
+		goglog.Logger.Error("topics should not be empty")
+		return nil, err
+	}
+	conf.saConf = sarConfig
+
+	conf.Codec, err = config.GetCodecOrDefault(ctx, *raw)
+	if err != nil {
+		return nil, err
+	}
+
+	return &conf, nil
+}
+
+// Start wraps the actual function starting the plugin
+func (t *InputConfig) Start(ctx context.Context, msgChan chan<- logevent.LogEvent) (err error) {
+	/**
+	 * Setup a new Sarama consumer group
+	 */
+	cum := consumerHandle{
+		i:     t,
+		ch:    msgChan,
+		ready: make(chan bool),
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	client, err := sarama.NewConsumerGroup(t.Brokers, t.Group, t.saConf)
+	if err != nil {
+		goglog.Logger.Errorf("Error creating consumer group client: %v", err)
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			// `Consume` should be called inside an infinite loop, when a
+			// server-side rebalance happens, the consumer session will need to be
+			// recreated to get the new claims
+			if err := client.Consume(ctx, t.Topics, &cum); err != nil {
+				goglog.Logger.Errorf("Error from consumer: %v", err)
+			}
+			// check if context was cancelled, signaling that the consumer should stop
+			if ctx.Err() != nil {
+				return
+			}
+
+			cum.ready = make(chan bool)
```

Comment:
Why renew the ready channel? (just curious)

Reply:
the example of sarama written like this,I think the reason is this. In the `Setup` func, it was originally written `close (c.ready)`, so need to renew the ready channel, but I changed it to `c.ready <-true` in the setup, and I will try to It changed back.

Reply:
I guess the example assumes you would reuse the channel. But in this case, the `Start()` executes only once. Anyway, renew the channel might not break anything.